Add a `gif me <search term>` responder

stevegrossi opened the pull request:
<img width="391" alt="screen shot 2016-10-02 at 11 44 58 am" src="https://cloud.githubusercontent.com/assets/664341/19021719/ae30366c-8895-11e6-9747-be26da9f261a.png">

Ultimately, this should probably be its own package (maybe for submission to [awesome-hedwig](https://github.com/enilsen16/awesome-hedwig#responders)!)

stevegrossi committed:
Test for regexes

stevegrossi committed:
`gif me` behavior to return a GIF URL for search term

Need to stub the API in tests

stevegrossi committed:
Tests with mock module

Hat tip: https://hashrocket.com/blog/posts/mocking-api-s-with-elixir

stevegrossi committed:
Revert "Tests with mock module"

This reverts commit 4b1552a15a2b171bff06e7b020b637e042ae8223.

stevegrossi committed:
Tests with ExVCR

stevegrossi committed:
Refactor

stevegrossi commented on test/responders/gif_me_test.exs:
```diff
@@ -0,0 +1,13 @@
+      assert String.ends_with?(text, ".gif")
+    end
+  end
+end
```
[`ex_vcr`](https://github.com/parroty/exvcr) (where `use_casette` comes from) is ultimately a mocking library, and José [famously opposes this kind of mock](https://twitter.com/josevalim/status/641617411242913792) in Elixir. See the commit https://github.com/indyelixir/elix/pull/1/commits/623cdc994f27b017a3a06b445b769898371a3143 for an  alternative which:
- extracts the HTTP request-making code into a separate module (`GiphyClient`)
- uses a different module in the test environment, which returns test data rather than make HTTP requests (`GiphyClientMock`—maybe not the right name since this is different than )

José recommends this approach in the Twitter convo above, and I followed a [great post from the Hashrocket folks](https://hashrocket.com/blog/posts/mocking-api-s-with-elixir) in doing this. However, it has some issues:
- if we extract the HTTP request-making logic into its own module, we'd _still_ need something like `ex_vcr` to test _that_, right?
- I don't love coupling this behavior to the environment, meaning you have to go look in `prod.exs` to ultimately figure out how `Responders.GifMe` does its job.

Still, who am I to disagree with the creator of Elixir? I'd like to use the `GiphyClientMock` approach instead of `ex_vcr`, but present both here for discussion.

janxious replied:
Very pro the `GiphyClientMock`. Having the robot be able to tell you when something goes wrong with your integration, either via external instrumentation or with just the error message as a response seems like a reasonable way to find out your integration is broken.

> if we extract the HTTP request-making logic into its own module, we'd _still_ need something like `ex_vcr` to test that, right?

Why is that? It seems like you have a known good case or cases, and you program for them, and in the case you meet a case you don't know about, crash and report?

stevegrossi replied:
I think you're right.

stevegrossi commented on lib/elix/responders/gif_me.ex:
```diff
@@ -0,0 +1,47 @@
+    |> build_api_url
+    |> make_request
+    |> get_random_image_url
+  end
```
This doesn't do any error handling. I'd like to use Elixir's `with` construct to handle errors in `make_request` (e.g. timeouts) and `get_random_image_url` (e.g. a different hash structure).

janxious replied:
So something like:

``` elixir
with {:ok, url}  <- build_api_url(search_term),
with {:ok, data} <- make_request(url),
     do: get_random_image_url(data)
else
  {:error, :bad_url}     -> {:error, "we couldn't process your search term"}
  {:error, :bad_request} -> {:error, "we had trouble talking to giphy"}
  {:error, :no_url}      -> {:error, "we couldn't find an image for that"}
  {:error, reason}       -> {:error, "something bad happened: #{reason}}
end
```

With the assumption that `get_random_image_url` will return `{:ok, url}`?

stevegrossi replied:
I ended up not needing `with` because with the extraction of `APIClient`, there's only one way this can fail (in `make_request`) and `handle_response\1` already handles that. I think `with` is when there are potential errors in multiple stages of a pipelines.

stevegrossi commented on lib/elix/responders/gif_me.ex:
```diff
@@ -0,0 +1,47 @@
+  defp get_random_image_url(response) do
+    Enum.random(response["data"])["images"]["original"]["url"]
+  end
+end
```
Hex.pm doesn't have a Giphy API client yet, but much of this logic can/should probably be extracted into a separate module, perhaps one that [wraps `HTTPoison.Base`](https://github.com/edgurgel/httpoison#wrapping-httpoisonbase) which has a nice API for such things.

janxious replied:
Perhaps shaving the yak first is a good idea? Make a nice giphy API client and then depend on it? :dancing_men:

stevegrossi replied:
Perhaps eventually, but I decided against an Elixir wrapper to the Giphy API since we only need a single GET endpoint right now.

janxious commented on fixture/vcr_cassettes/giphy_search_steve.json:
```diff
@@ -0,0 +1,35 @@
+      "url": "http://api.giphy.com/v1/gifs/search?q=gif%20me%20steve&api_key=dc6zaTOxFJmzC&limit=10"
+    },
+    "response": {
+      "body": "\n\n{\"data\":[{\"type\":\"gif\",\"id\":\"4ah61Apr8oAEw\",\"slug\":\"steve-martin-4ah61Apr8oAEw\",\"url\":\"http:\\/\\/giphy.com\\/gifs\\/steve-martin-4ah61Apr8oAEw\",\"bitly_gif_url\":\"http:\\/\\/gph.is\\/2c5JRU5\",\"bitly_url\":\"http:\\/\\/gph.is\\/2c5JRU5\",\"embed_url\":\"http:\\/\\/giphy.com\\/embed\\/4ah61Apr8oAEw\",\"username\":\"\",\"source\":\"http:\\/\\/www.funnyjunk.co ...
```
I wonder if this could be broken into multiple lines just for possible future human understanding, or maybe it's not for humans so it don't matter.

stevegrossi replied:
It's auto-generated by `ex_vcr` so, yes _until_ we need to change it, then we'll have to break it up again. To me, feels like "not for humans"... It may become a non-issue when I build my own mock object.

janxious replied:
![gif-keyboard-17288769395354051567](https://cloud.githubusercontent.com/assets/50124/19050318/11037c18-897c-11e6-8389-d443841762e5.gif)

janxious commented on mix.exs:
```diff
@@ -30,7 +30,9 @@ defmodule Elix.Mixfile do
       {:hedwig, github: "hedwig-im/hedwig"},
-      {:hedwig_flowdock, "~> 0.1"}
+      {:hedwig_flowdock, "~> 0.1"},
+      {:httpoison, "~> 0.9.0"},
```
I looked briefly at how hedwig handles HTTP and it was not super involved. `httpoison` is definitely easier, but it also seems like it adds a _lot_ of dependencies, but I guess that could also be `exvcr`

janxious commented on lib/elix/responders/gif_me.ex:
```diff
@@ -0,0 +1,47 @@
+  @usage """
+  gif me <term> - Replies with a GIF URL matching the term
+  """
+  hear ~r/gif me (.+)/i, msg do
```
Perhaps this should be `\A\s+gif me…`?

What is the pre-processing done on the string before it gets here?

stevegrossi replied:
Changing this to `(.*)`, it looks like what we're matching on is the full text of what the user typed, which means it doesn't currently matter whether or not you do "@elix gif me something" or just "gif me something" or "who wants some GIF meat?" I'm inclined to make the `@elix` bit optional, but I definitely want to add some word boundaries. Gotta watch out for that GIF meat...

stevegrossi replied:
Nevermind, this won't match "want some gif meat", but it will match "he tried to punch.gif me in the face!" which, on second thought, might actually have comedic value...

janxious commented:
Best feature of any bot is streaming gifs.

stevegrossi committed:
Whoops, actually it’s the second match

The first (`[0]`) is the entire string. IO.inspect is your friend, folks.

stevegrossi committed:
Revert "Tests with ExVCR"

This reverts commit ceed3d72cb5290aac9fd2950a7d542db289d0a4b.

stevegrossi committed:
Tests with mock module

Hat tip: https://hashrocket.com/blog/posts/mocking-api-s-with-elixir

stevegrossi committed:
Merge branch 'master' into gif-me

stevegrossi commented on config/test.exs:
```diff
@@ -2,3 +2,5 @@ use Mix.Config
 config :elix, Elix.Robot,
   adapter: Hedwig.Adapters.Test
+
+config :elix, :giphy_client, Elix.GiphyClientMock
```
Hm, I get a compile warning when running `mix test` that `Elix.GiphyClientMock` is not available:

``` elixir
warning: function Elix.GiphyClientMock.get/1 is undefined (module Elix.GiphyClientMock is not available)
  lib/elix/responders/gif_me.ex:41
```

The tests still pass because I manually `require` the file in `test_helper.exs`, but I'd like to avoid the compiler warning by making this file available at compile time. (Changing the extension to `.exs` doesn't avoid the compile warning. I should figure out how to do that.

stevegrossi replied:
Ah, [this SO comment](http://stackoverflow.com/questions/30652439/importing-test-code-in-elixir-unit-test#comment49371142_30652675) (Thanks José!) pointed me to [how Phoenix requires and compiles `test/support`](https://github.com/phoenixframework/phoenix/blob/db2f953f24973fab8aa2aab3bd8fc9c87a17adc8/installer/templates/new/mix.exs#L10).

stevegrossi committed:
Fix compilation warning for spec/support

Ah, [this SO comment](http://stackoverflow.com/questions/30652439/importing-test-code-in-elixir-unit-test#comment49371142_30652675) (Thanks José!) pointed me to [how Phoenix requires and compiles `test/support`](https://github.com/phoenixframework/phoenix/blob/db2f953f24973fab8aa2aab3bd8fc9c87a17adc8/installer/templates/new/mix.exs#L10).

stevegrossi committed:
Generalize APIClient

Come to think of it, there’s nothing really Giphy specific in here

stevegrossi committed:
Better error handling

janxious commented on lib/elix/api_client.ex:
```diff
@@ -0,0 +1,14 @@
+    |> handle_response()
+  end
+
+  defp handle_response({:ok, response}) do
```
How do you feel about one-lining these `handle_response`s matchers?

stevegrossi replied:
👍

stevegrossi replied:
Done! Though I added a couple `alias`es to keep these lines under 80 characters.

janxious commented on lib/elix/responders/gif_me.ex:
```diff
@@ -0,0 +1,50 @@
+  @usage """
+  gif me <term> - Replies with a GIF URL matching the term
+  """
+  hear ~r/gif me (.+)/i, msg do
```
Worth adding in the [`from_self`](https://github.com/indyelixir/elix/commit/b10e628ca07fcc0406c2d0eb845f8b18d375fcdc#diff-ffe52e6014b7809b6bfc8713165e838aR33) bits here? Seems unlikely elix will be talking to itself about gifs, but worth a thought. I relaly wonder why `ignore_self`  was removed

stevegrossi replied:
I decided against it in this PR only because it's impossible for Elix to return something containing "gif me" (it'll be either a URL or the "I'm sorry..." message). Ultimately, I do think we'll want this filter applied globally, but I'd prefer to tackle that separately in a better way.

stevegrossi replied:
I posed this question to the hedwig issue tracker: https://github.com/hedwig-im/hedwig/issues/57

stevegrossi committed:
Address feedback and add some documentation

stevegrossi committed:
Merge branch 'master' into gif-me

stevegrossi commented:
Thanks for the great feedback, @janxious! Any final thoughts before I merge?

janxious commented:
> Any final thoughts before I merge?

:shipit:
